sstables/partition_index_cache.hh

```diff
@@ -246,21 +246,43 @@ public:
     // Evicts all unreferenced entries.
     future<> evict_gently() {
         auto i = _cache.begin();
-        while (i != _cache.end()) {
+        std::optional<partition_index_page> partial_page;
+        auto clear_on_exception = defer([&] {
+            with_allocator(_region.allocator(), [&] {
+                partial_page.reset();
+            });
+        });
+        while (partial_page || i != _cache.end()) {
+          if (partial_page) {
+            auto preempted = with_allocator(_region.allocator(), [&] {
+                while (!partial_page->empty()) {
+                    partial_page->clear_one_entry();
+                    if (need_preempt()) {
+                        return true;
+                    }
+                }
+                partial_page.reset();
+                return false;
+            });
+            if (preempted) {
+                auto key = i->key();
```

Comment:
I think it's possible that i == _cache.end() here, in which case this is UB.

Reply:
Good catch.